File: leetcode/0257-binary-tree-paths/solution.py
from typing import List, Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        paths = []

        def dfs(node: Optional[TreeNode], path: str) -> None:
            if node is None:
                return

            current = path + str(node.val) if not path else path + "->" + str(node.val)

            if node.left is None and node.right is None:
                paths.append(current)
                return

            dfs(node.left, current)
            dfs(node.right, current)

        dfs(root, "")
        return paths

File: leetcode/0257-binary-tree-paths/solution.cpp
#include <string>
#include <vector>
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> paths;
        dfs(root, "", paths);
        return paths;
    }

private:
    void dfs(TreeNode* node, string path, vector<string>& paths) {
        if (!node) {
            return;
        }

        string current = path.empty() ? to_string(node->val) : path + "->" + to_string(node->val);

        if (!node->left && !node->right) {
            paths.push_back(current);
            return;
        }

        dfs(node->left, current, paths);
        dfs(node->right, current, paths);
    }
};
